### ISSUE TITLE
Maximize function does not resize to fullscreen in certain cases

### ISSUE THREAD

**Issue by preetsethi** [rxhanson/Rectangle#15]
Depending on window placement (and possibly size of the window?), when I trigger the _**Maximize**_ function, there are certain cases where it resizes or moves the window, but does not resize the window to the maximum size allowed on the first call. In the end, I am able to chain together a few _**Maximize**_ calls to get the desired effect, usually requiring 3 calls (or pressing the hotkey 3 times in a row).

Here's some screenshots to demonstrate what I am talking about.

1. Starting window placement (I used Rectangle's _**Bottom Left**_)
<img width="1680" alt="Screen Shot 2019-09-27 at 6 39 09 AM" src="https://user-images.githubusercontent.com/8645918/65774385-fb0ab080-e0f2-11e9-832f-ce73a6d1c038.png">

2. After I trigger _**Maximize**_ (first invocation):
<img width="1680" alt="Screen Shot 2019-09-27 at 6 39 22 AM" src="https://user-images.githubusercontent.com/8645918/65774308-d1ea2000-e0f2-11e9-8f93-1a226ccfeb0a.png">

3. After I trigger _**Maximize**_ (second time):
<img width="1680" alt="Screen Shot 2019-09-27 at 6 39 35 AM" src="https://user-images.githubusercontent.com/8645918/65774361-ef1eee80-e0f2-11e9-9dc4-ef934c699137.png">

4. After I trigger _**Maximize**_ (third time) I finally get the desired result:
<img width="1680" alt="Screen Shot 2019-09-27 at 6 39 44 AM" src="https://user-images.githubusercontent.com/8645918/65774464-27263180-e0f3-11e9-9320-ac638568ed2b.png">

As I said, this only happens in certain cases. If I use _**Top Left**_ for some window, and then I try to _**Maximize**_, I get the desired effect on the first call.

These are the only cases I can confirm are 100% reproducible for me. However, I see the issue quite a bit with various window placements / sizes. When I start the window in the _**Bottom Right**_ placement, I get annoyingly variable results when triggering _**Maximize**_ next. Sometimes it will succeed on the first call but more often it will require two calls.

I am not running any external monitors, I experience these issues only using the built-in display on my Macbook Pro. I don't run multi-monitor at all these days, but I can easily set this up and test under that setup as well. 

So to provide some more background (and probably useless info):
1. Before Rectangle, I used Spectacle. Spectacle had the same  _**Maximize**_ function. I first saw this issue with Spectacle. I am assuming that some Mac OS update broke this, but I sadly can't provide more details here. I finally got annoyed enough that I decided to at least try to fix this. 
2. When I went looking through the open and closed issues in the Spectacle repo, I found a lot of info, and _I THINK I REMEMBER_ finding some comments detailing the mechanism in which Spectacle resizes windows and that some Mac OS update had changed this? I was planning on filing an issue detailing my experience and referencing those comments, but almost immediately after, I found your comment introducing Rectangle in another issue. I lost focus, downloaded Rectangle, forgot what I was doing originally, and closed that browser window with all the tabs. 

As I said, that last bit was probably useless info, haha.

---

I would sincerely appreciate you looking into this! 
As always, willing to provide any additional info you need in case this is not easily reproducible.

**Comment by rxhanson** [comment]
This might take a logging build to figure out, as I can't reproduce it. I'm going to try a few more things to see if I can reproduce it, and I'll search through the Spectacle issues some more as well.

Are you using the default keyboard shortcuts, or have you changed any of them? It's some pretty strange behavior, and off the top of my head I can't think of what would cause this. Window size/location shouldn't affect the maximize behavior because it's just setting the window size/location to the visible frame of the screen, which is obtained from Apple's code. Having just the one screen simplifies this even more. If you use the menu from the menu bar to execute the window action, does it also have the same behavior?

**Comment by preetsethi** [comment]
I am using custom shortcuts:
<img width="747" alt="Screen Shot 2019-10-02 at 1 52 34 AM" src="https://user-images.githubusercontent.com/8645918/66030999-62d84700-e4b7-11e9-8f1c-a9815b259e15.png">

Using the menu from the menu bar has the same effect, consistent behavior to what I described above, I have to hit _**Maximize**_ three times from the menubar applet to get the desired window placement and size.

The fact that I am seeing this behavior across Spectacle <->  Rectangle, this fully puts me in the camp that I am doing something horribly wrong lol, just trying to figure out what that is. This is a dev machine, so I definitely could have some strange interaction going on with some other library or utility app I installed haphazardly -- but I cannot think of a single thing I've installed that would interfere with window placement / resize logic. I mainly do web dev ... so I've been confused to say the least.

I would gladly take up the task of getting a logging build going and getting you any information you need. I'm gonna start by building from source and getting that going, but if you can provide even the most basic instructions in terms of what you mean by `logging build`, I would sincerely appreciate that.

**Comment by rxhanson** [comment]
It could be a bug that got carried from Spectacle. At a start, I've created the branch `log-action` (https://github.com/rxhanson/Rectangle/tree/log-action) where I added a log message that might tell us what's going on. It logs the action being executed, the display frame, the calculated frame of the window that we are attempting to set, and the resulting frame of the window after attempting to set it.

It's definitely easiest on my end if you're able to run Rectangle from the source and see the log messages come in. Since it uses CocoaPods, you'll have to run a pod install in the root. If you run into issues, let me know. If it comes down to it, I can turn that branch into a build that you can download and try.

**Comment by preetsethi** [comment]
@rxhanson 
Going to work on building this from source first, it'll be fun to mess around with something new. Hope to find some time to do this during the weekend, will report back.

**Comment by rxhanson** [comment]
@preetsethi 
I finally got around to adding debug logging in to a release. Download it at https://github.com/rxhanson/Rectangle/releases/tag/v0.9.1 

From the readme:

> ## Debug Logging
> To enable debug logging (starting in v0.9.1), execute the following Terminal command and restart the app:
> 
> `defaults write com.knollsoft.Rectangle debug -bool true`
> 
> To view logs:
> 1. Open Console.app
> 1. Select Action -> Include Debug Messages in the menu
> 1. Enter this in the search bar: `process:Rectangle any:calculatedRect`
> 

I'll be curious to know what the logs show.

**Comment by preetsethi** [comment]
@rxhanson 
I really appreciate this update! Seriously, thanks man. 
Downloading it right now, will install and get back to you about my original issue promptly.

---

Side note: I did attempt to build the app! Installed cocoapods, navigated through some interesting turns with ruby installed with `rvm` vs sys ruby. I got hung up and halted by build issues in xcode that I believe center around code signing: I am _not_ a member of the Apple Developer Program. So just wanted to naively ask this to validate my conclusion:
1. Do you need to be a member of the Apple Developer Program to build Rectangle (or at least the way it's configured to build out of the box)?

**Comment by rxhanson** [comment]
I'll plan on adding a quick how to build section in the readme. 

> Do you need to be a member of the Apple Developer Program to build Rectangle (or at least the way it's configured to build out of the box)?

The short answer is usually not. Since I have to sign Rectangle, the project configuration for a release is typically left in a state where it's ready to be archived and submitted to Apple for notarization. Often, I try to make sure that I revert it back to not having my developer id in the signing config, and have it set to run locally (but sometimes I forget - sorry!). Otherwise, you'd want to go in and manually select sign to run locally for each part of the project (Rectangle, RectangleLauncher & tests), and I think you'd need to remove Sandbox and hardening from each of those as well. In that state, it should be fine to just hit run, but every time you run when signing to run locally, you'll have to disable and re-enable Rectangle in the Privacy/Security System Prefs. I think you can set up your own local developer id and use that instead of signing to run locally, but I haven't tried that in a long time. 

For this past release, I did leave it in the "Sign to run locally" state, so you should be able to just clone it, install cocoapods, and run it. 

Thanks for trying to run it though! Good to know so that I can improve it.

**Comment by preetsethi** [comment]
Appreciate the quick response man! Glad to know I didn't miss something obvious. I'll try building again sometime with the most recent version.

But to report back, here's the debug logging from a similar window placement and size I described in my original post when I opened this issue (see note): 
```
Rectangle maximize | display: (0.0, 0.0, 1680.0, 1027.0), calculatedRect: (0.0, 23.0, 1680.0, 1027.0), resultRect: (-0.0, 506.0, 1680.0, 544.0)
Rectangle maximize | display: (0.0, 0.0, 1680.0, 1027.0), calculatedRect: (0.0, 23.0, 1680.0, 1027.0), resultRect: (0.0, 23.0, 1680.0, 544.0)
Rectangle maximize | display: (0.0, 0.0, 1680.0, 1027.0), calculatedRect: (0.0, 23.0, 1680.0, 1027.0), resultRect: (-0.0, 23.0, 1680.0, 1027.0)
```
Note: My original post used the _**Bottom Left**_ Rectangle action to setup window placement. A sad development is that the _**Bottom Left**_  and _**Bottom Right**_ actions no longer work in `0.9.1`. The last version I had installed was `0.7` where those actions worked (_if memory serves me right_). 

---
Here's some debug logging regarding _**Bottom Left**_ for example:
Starting window placement is a fully maximized window.
```
Rectangle	bottomLeft | display: (0.0, 0.0, 1680.0, 1027.0), calculatedRect: (0.0, 537.0, 840.0, 513.0), resultRect: (-0.0, 23.0, 840.0, 513.0)
Rectangle	bottomLeft | display: (0.0, 0.0, 1680.0, 1027.0), calculatedRect: (0.0, 537.0, 1120.0, 513.0), resultRect: (-0.0, 23.0, 1120.0, 513.0)
Rectangle	bottomLeft | display: (0.0, 0.0, 1680.0, 1027.0), calculatedRect: (0.0, 537.0, 560.0, 513.0), resultRect: (-0.0, 23.0, 560.0, 513.0)
Rectangle	bottomLeft | display: (0.0, 0.0, 1680.0, 1027.0), calculatedRect: (0.0, 537.0, 840.0, 513.0), resultRect: (-0.0, 23.0, 840.0, 513.0)
```
So the first call to the action results in the window occupying the top-left quadrant of the screen.
The subsequent calls result in the expected `1/3 - 2/3 - 3/3` window resizing in that frame position.

---
SO...
Using your debug output, I noticed the `resultRect` / `resultingRect` had an odd `-0.0` x-cord origin for the window in some case. From my vast experience f***iing up floating point calculations in my code, I knew something was up. I proceed to go down a _deep_ rabbit hole.

Sadly: I have nothing to report except confusion.
But I'm gonna summarize my confusion and hopefully that helps you move forward.

tl;dr - I think there's some funky stuff going on between `NSRect` <-> `CGRect` <-> `AXUIElement`

Like I said, I went down a deep rabbit hole exploring Apple developer APIs / StackOverflow / Github issues, purely out of my own interest, but the thing that stuck was the difference between origin of `(x,y)` coordinate definitions. In some Apple APIs it's top-left, and in others its bottom-left. Perhaps this is all a red herring and there's a completely different issue. But here's some links I found interesting:

https://github.com/tmandry/Swindler/issues/62
Note: Swindler looks cool!
https://github.com/tmandry/Swindler/blob/5c1d49d407cbd7066e09f4cfab3e26f7cf3a8ba8/Sources/Window.swift#L59
```
/// The origin of the frame is the bottom-left corner of the window in screen coordinates.
```
That led me to this:
https://github.com/tmandry/Swindler/blob/5c1d49d407cbd7066e09f4cfab3e26f7cf3a8ba8/Sources/Window.swift#L319
```
/// This is to convert between AXPosition coordinates, which have the origin at
/// the top-left, and Cocoa coordinates, which have it at the bottom-left.
```
It's completely possible you're accounting for this and I missed that in the Rectangle code. Like I said, I went down a rabbit hole. Excuse my novice opinion if I'm completely off base, but it was a fun exploring this code nonetheless!

In summary, let me know if you need anything else from me! I'm gonna try to build the most recent build in the coming week sometime. But please reach out ASAP if you have more cases that you want me to test.

--- 

Personal Note:
I really appreciate you following up on this issue man. Your activity and support, they further solidify my opinion: I was f***ing up not supporting my favorite free software. Never again.

**Comment by rxhanson** [comment]
Thanks for reporting back! I'll dig into it deeper when I get some more time. Until then, which version of macOS are you on?

**Comment by preetsethi** [comment]
@rxhanson 
`10.14.6 (18G95)`
<img width="698" alt="Screen Shot 2019-10-27 at 4 01 51 AM" src="https://user-images.githubusercontent.com/8645918/67633576-be0b1880-f86e-11e9-9a05-0b20df934931.png">

**Comment by preetsethi** [comment]
@rxhanson I've been thinking about reproducibility this whole time, and something finally clicked: 
I have the Accessibility Keyboard enabled. Any corner of my screen is configured as a hot-corner to display/hide the on-screen keyboard.

![image](https://user-images.githubusercontent.com/8645918/67919189-9660cc80-fb5c-11e9-9a3c-ff9ebe74722e.png)

The reasoning of enabling this is _extreme laziness_: when I'm in bed watching Netflix and the touchpad is more accessible than the keyboard, I can move my mouse to any corner of the screen and pop open the Accessibility Keyboard, adjust Brightness of Volume for example.
 
**However, I have tried disabling the Accessibility Keyboard (and restarting), yet I still have the same results.**

I first and foremost wanted to mention that ^
Perhaps try enabling Accessibility Keyboard on your machine?

---

@rxhanson Listen, you bring up something pertinent: no one else is having this issue.
**I appreciate your responses and triage so far.**
In my previous issue, you mentioned I should detail my behavior, so I felt obligated.
Frankly dude, I am the first to admit this is not an issue thats easily resolved. 
I've had experience as a production support engineer, and my favorite product manager always joked about engineers that threw out the "works on my machine" cop-out haha. This is the opposite case: breaks on my machine haha.

**Feel free to handle this as you see fit, do not feel obligated in any manner, but I will be here in any case to respond, support and advocate. 
You've been freaking awesome, keep it up man.**

**Comment by rxhanson** [comment]
Thanks for gutting through the thorough Accessibility testing, and glad you got the build working locally! That really could accelerate getting to the bottom of this. 

I didn't notice that you had the build working locally, so I went ahead and published a release that removes the `BestEffortWindowMover` instance in the `WindowManager`. Testing locally, things worked for me without it (but they didn't when I had previously tried this months ago - I'm on Catalina now). You can grab the release at https://github.com/rxhanson/Rectangle/releases/tag/v0.9.1.a. Or, you can go in and remove the `BestEffortWindowMover` and run it locally. 

Let me know what happens. If you have identical behavior before and after removing it, then we can at least rule it out as the cause and look directly at `AccessibilityElement.normalizeCoordinatesOf(...)`. If things are further broken, then we'll know that it is essential (at least on Mojave).

If we rule out both the `BestEffortWindowMover` and `AccessibilityElement.normalizeCoordinatesOf(...)`, then my current thought is that it might be an obscure Apple bug that we might not be able to fix. Really then what this boils down to is:
Do the following two Apple Accessibility functions do what they're supposed to?
`AXUIElementSetAttributeValue(self.underlyingElement, kAXPositionAttribute as CFString, value)`
`AXUIElementSetAttributeValue(self.underlyingElement, kAXSizeAttribute as CFString, value)`

**Comment by preetsethi** [comment]
@rxhanson 
### tl;dr -> **I am closing this issue! Had nothing to do with Rectangle specifically.** 

**BUT** your support was fundamental in helping me root out the problem. 
**Seriously!!!** -> You are awesome, thanks for your help! Your response to this issue helped me do the following:
1. You helped me learn! In the process of setting up and debugging an XCode environment that was able to build and test changes to your app: Swift is cool. I understand Swift. I can debug Swift. When I explored Objective-C ages ago, I found some obtuse principles that made the language unattractive to a developer with my pedigree and past experience. I am going to keep exploring Swift and learn. THIS IS AN AWESOME THING, and your support in this thread is fundamentally responsible for that.
1. Your addition of debug logging helped me figure something out: **fundamentally, the calculations were correct**. My previous post pointed at some intricacies and lack of clarity regarding various Apple APIs exposed to handle window management. I explored these more. I explored the different `WindowMover` classes as you recommended. All of this pointed at the fact that the problem was outside of Rectangle (or Spectacle, as both apps had the same issue).

### the murky details
I'm not 100% sure of my findings as we speak, but two things helped me resolve my issues:
1. I realized iTerm2 had some conflicting Keyboard Shortcut bindings. I had given iTerm2 Accessibility Access in the past, but I had unchecked it, thinking that this would eliminate any confounding behavior between iTerm2 <-> Rectangle. Finding this conflicting shortcut, even though I had eliminated iTerm2's ability to use accessibility features... I removed the application altogether from the System Pref's pane. **THIS ALONE DID NOT SOLVE MY PROBLEM**
1. I started to look into how OS X stored global keyboard shortcuts. I found a lot of confounding answers regarding various system plists, stored all over the place. I'm a little mad at myself for not being more methodical about this but... after trying to _reset_ ALL of the various plists I found in even _loosely related_ stackoverflow / superuser / Apple discussions threads... somewhere along the line, I reset the behavior and achieved expected results from Rectangle. Some threads suggested that its possible these plists were corrupted along the way? I'm not going to comment on the validity of that specifically, but wanted to share nonetheless.

@rxhanson Dude. I hate that you had to go through this with me to debug an environment-specific / user-specific issue. As someone with both IT and software support experience, I am sorry! I looked specifically at the software and not the environment first. That's my bad! **HOWEVER...**

I still really appreciate this experience and thank you for your active support and development into this open-source project. What you're doing is cool. You're cool. Thanks man!

**Comment by rxhanson** [comment]
Woohoo! Glad the problem is solved. 

I really enjoy writing Swift, so I’m glad this has been a positive experience for you as well. 

Thanks for sharing about the shortcuts, too - those are actually valuable tidbits of information that we wouldn’t get if we hadn’t gone this far down this path. 

Let me know if anything else comes up!

**Comment by ospfranco** [comment]
@preetsethi Hey, I'm having the same issue I have to press the global shortcut a bunch of time until the window places itself correctly, could you mind point where exactly you reseted the plists to get it working properly? thanks!

**Comment by rxhanson** [comment]
@ospfranco, there are no animations with Rectangle resizing, so this would point to another app or something else causing this. 

Also, some users with similar issues reported the behavior went away when they updated to the latest version of macOS. As always, it's hard to tell what was actually going on with users machines in instances like this.

**Comment by ospfranco** [comment]
I also disabled intern accessibility access and now everything seems to work, maybe it was that? I checked Alfred and there is no window management funcionality

**Comment by preetsethi** [comment]
@ospfranco 
Sorry for the late post. 
Deleting the keyboard plist turned out to be useless, did not resolve my issue on its own. 

My original problem was with iTerm2. If I give that app Accessibility Access, iTerm2 interferes with Rectangle, even if there are no conflicting keyboard shortcut assignments across the apps. 
I too, use Alfred. Since my earlier posts, I had this problem re-appear with Alfred when I gave it accessibility access.

**Conclusion: _only give Rectangle accessibility access_.**

Possible next steps: knowing the wonky behavior appears when enabling accessibility access for iTerm2 / Alfred ... at least that's reproducible and someone could begin to debug this. Frankly, neither iTerm2 nor Alfred really need Accessibility Access for my use cases.

Good luck!

**Comment by cotfas** [comment]
@preetsethi 

> The reasoning of enabling this is _extreme laziness_: when I'm in bed watching Netflix and the touchpad is more accessible than the keyboard, I can move my mouse to any corner of the screen and pop open the Accessibility Keyboard, adjust Brightness of Volume for example.

Had the same issue and the Multitouch app helps

<img width="277" alt="image" src="https://user-images.githubusercontent.com/3340367/79641323-7d49e780-819f-11ea-88a3-0ecdfdd5bf3e.png">

**Comment by ospfranco** [comment]
Yeah, I can confirm that when neither iTerm 2 or Alfred have the accessibility features turned on, Rectangle works perfectly, should be useful information to further debug this issue

**Comment by rxhanson** [comment]
This should now be fixed in [v0.40](https://github.com/rxhanson/Rectangle/releases/tag/v0.40).
